url542ae48576_tonyheadford_m68k_tgz-pJ8-m68k_MonitorJ8/src/m68k/cpu/instructions/MOVEM.java

```diff
@@ -0,0 +1,692 @@
+
+		// true up if (An)+ addressing mode
+		if(((opcode >> 3) & 0x07) == 3)
+		{
+			count = getMultipleWordPostInc(opcode & 0x07, reglist, address);
+		}
+		else
+		{
+			count = getMultipleWord(reglist, address);
+		}
+
+		return M2R_Timing[src.index()] + (count << 2);	//reg count * 4
+	}
+
+	protected final int movem_long_m2r(int opcode)
+	{
+		int reglist = cpu.fetchPCWord();
+		Operand src = cpu.resolveSrcEA((opcode >> 3) & 0x07, opcode & 0x07, Size.Long);
+		int address = src.getComputedAddress();
+		int count;
+
+		//true up if (An)+ addressing mode
+		if(((opcode >> 3) & 0x07) == 3)
+		{
+			count = getMultipleLongPostInc(opcode & 0x07, reglist, address);
+		}
+		else
+		{
+			count = getMultipleLong(reglist, address);
+		}
+
+		return M2R_Timing[src.index()] + (count << 3);	//reg count * 8
+	}
+
+	public final DisassembledInstruction disassembleOp(int address, int opcode, Size sz, boolean reg_to_mem)
+	{
+		DisassembledOperand src;
+		DisassembledOperand dst;
+		int mode = (opcode >> 3) & 0x07;
+		int reg = (opcode & 0x07);
+		int reglist = cpu.readMemoryWord(address + 2);
+		boolean reversed = (mode == 4);
+
+		if(reg_to_mem)
+		{
+			//registers to memory
+			src = new DisassembledOperand(regListToString(reglist, reversed), 2, reglist);
+			dst = cpu.disassembleDstEA(address + 4, mode, reg, sz);
+		}
+		else
+		{
+			// memory to registers
+			src = cpu.disassembleSrcEA(address + 4, mode, reg, sz);
+			dst = new DisassembledOperand(regListToString(reglist, reversed), 2, reglist);
+		}
+
+		return new DisassembledInstruction(address, opcode, "movem" + sz.ext(), src, dst);
+	}
+
+	protected final String regListToString(int reglist, boolean reversed)
+	{
+		StringBuilder sb = new StringBuilder();
+		int first = -1;
+		int count = 0;
+
+		if(!reversed)
+		{
+			//normal mode lsb = d0
+			char prefix = 'd';
+			int mask = 1;
+
+			for(int i = 0; i < 2; i++)
+			{
+				for(int n = 0; n < 8; n++, mask <<= 1)
+				{
+					if((reglist & mask) != 0)
+					{
+						if(first != -1)
+						{
+							count++;
+						}
+						else
+						{
+							first = n;
+						}
+					}
+					else
+					{
+						if(first != -1)
+						{
+							if(sb.length() > 0)
+								sb.append('/');
+
+							sb.append(prefix);
+							sb.append(first);
+							if(count == 1)
+							{
+								sb.append('/');
+								sb.append(prefix);
+								sb.append(n - 1);
+							}
+							else if(count > 1)
+							{
+								sb.append('-');
+								sb.append(prefix);
+								sb.append(n - 1);
+							}
+
+							count = 0;
+							first = -1;
+						}
+					}
+				}
+
+				if(first != -1)
+				{
+					if(sb.length() > 0)
+						sb.append('/');
+
+					sb.append(prefix);
+					sb.append(first);
+					if(count == 1)
+					{
+						sb.append('/');
+						sb.append(prefix);
+						sb.append(7);
+					}
+					else if(count > 1)
+					{
+						sb.append('-');
+						sb.append(prefix);
+						sb.append(7);
+					}
+
+					count = 0;
+					first = -1;
+				}
+
+				prefix = 'a';
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 268:336, 344:412.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.